src/card_number/mod.rs

```diff
@@ -1,42 +1,46 @@
-pub fn verify_card_number<T: AsRef<str>>(number: T) -> bool {
-    let number = number.as_ref();
-    let len = number.len();
-    if len < 16
-        || number.chars().any(|c| !c.is_numeric())
-        || number[1..10].parse::<u32>().unwrap() == 0
-        || number[10..16].parse::<u32>().unwrap() == 0
+trait CardVerification {
+    fn is_valid_card_number(&self) -> bool
+    where
+        Self: AsRef<str>
```

Comment:
Why is the whole trait not bounded by this? 
```suggestion
trait CardVerification: AsRef<str> {
    fn is_valid_card_number(&self) -> bool 
```

Now you allow `CardVerification ` to be implemented for _anything_, and I doubt if that makes any sense.

Reply:
Agree, yours makes absolutely more sense.

Reply:
Then it's not possible easily: https://github.com/rustland-fa/persian-tools-rs/pull/6#issuecomment-787093145
Which is not a big deal IMO

Reply:
Even before, if you'd implement `CardVerification` for `Vec<u8>` I think then you'd have to implement the method again since the trait bounds thereof were not satisfied, which brings me to the exact reason why it had to change. 

I think devs will have an easy time converting their `Vec<u8>` or `&[u8]` to str and pass it in. 

If you really want to be versatile, and support all, I think you need a wrapper trait that tries to convert any of `&str`, `String`, `&[u8]` and `Vec<u8>` to a `&str`. The problem would be that the latter two might fail and you need to deal with results. 

Something like: 

```
trait AsString {
   type Error;
   fn as_string() -> Result<&str, Self::Error>
} 

impl AsString for &str { .. }
impl AsString for Vec<u8> { .. }
impl AsString for &str { .. }
impl AsString for String { .. }

/// and finally
trait CardVerification: AsString { ... } 
```

But this is really an over-engineering example :D let the downstream user do the conversion. Supporting string is enough.

Reply:
Perfect. I agree :)